File: lz_encoder/encoder.py
class LZ77Tuple:
    def __init__(self, offset, length, char):
        self.offset = offset
        self.length = length
        self.char = char

    def __repr__(self):
        return f"({self.offset}, {self.length}, {repr(self.char)})"


def max_of_two(a : int, b : int) -> int:
    return a if a > b else b


def compress_lz77(history : str, lookahead : str) -> list[LZ77Tuple]:
    compressed_data = []
    window_size = len(history)
    i = 0

    while i < len(lookahead):
        best_match_distance = 0
        best_match_length = 0

        # Search for the longest match in the sliding window (history)
        for j in range(max_of_two(0, len(history) - window_size), len(history)):
            length = 0
            while length < len(lookahead) - i and history[j + length] == lookahead[i + length]:
                length += 1
            if length > best_match_length:
                best_match_distance = len(history) - j
                best_match_length = length

        # If a match is found, add a tuple (offset, length, next char)
        if best_match_length > 0 and i + best_match_length < len(lookahead):
            next_char = lookahead[i + best_match_length]
            compressed_data.append(LZ77Tuple(best_match_distance, best_match_length, next_char))
            i += best_match_length + 1
        else:
            # No match found, add a tuple with zero offset and length, and the current character
            compressed_data.append(LZ77Tuple(0, 0, lookahead[i]))
            i += 1

        # Update the history window
        if i < len(lookahead):
            history += lookahead[:i]
            if len(history) > window_size:
                history = history[len(history) - window_size:]

    return compressed_data


def slice_and_compress(input_str : str, n : int) -> None:
    # Ensure n does not exceed the length of input
    if n > len(input_str):
        n = len(input_str)

    new_text = input_str[:n]
    print(f"NewText: {new_text}")

    # Calculate the 2/3 and 1/3 split
    two_thirds_index = (2 * n) // 3
    history = new_text[:two_thirds_index]
    lookahead = new_text[two_thirds_index:]

    print(f"History: {history}")
    print(f"Lookahead: {lookahead}")

    # Apply LZ77 compression on history and lookahead
    compressed_data = compress_lz77(history, lookahead)

    # Print compressed data
    print("Compressed Data:")
    for tuple_ in compressed_data:
        print(tuple_)

    # Calculate memory size of compressed data
    memory_size_compressed = 0
    for tuple_ in compressed_data:
        memory_size_compressed += len(f"{tuple_.offset}{tuple_.length}{tuple_.char}")
    memory_size_original = len(new_text)

    print(f"Memory Size of Compressed Data: {memory_size_compressed}")
    print(f"Memory Size of Original Data: {memory_size_original}")

    # Compression ratio
    compression_ratio = memory_size_compressed / memory_size_original
    print(f"Compression Ratio: {compression_ratio:.2f}")


def run():
    input_str = "In radio c'è un pulcino In radio c'è un pulcino È il pulcino pio, è il pulcino pio E il pulcino pio, e il pulcino pio"
    plaque = "2077518"

    n = sum(int(digit) for digit in plaque)  # Sum of digits in plaque
    slice_and_compress(input_str, n)


if __name__ == "__main__":
    run()

File: lz_encoder/encoder.go
package main

import (
	"fmt"
)

type LZ77Tuple struct {
	offset int
	length int
	char   byte
}

func maxOfTwo(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func compressLZ77(history, lookahead string) []LZ77Tuple {
	var compressedData []LZ77Tuple
	windowSize := len(history)
	i := 0

	for i < len(lookahead) {
		bestMatchDistance := 0
		bestMatchLength := 0

		// Search for the longest match in the sliding window (history)
		for j := maxOfTwo(0, len(history)-windowSize); j < len(history); j++ {
			length := 0
			for length < len(lookahead)-i && history[j+length] == lookahead[i+length] {
				length++
			}
			if length > bestMatchLength {
				bestMatchDistance = len(history) - j
				bestMatchLength = length
			}
		}

		// If a match is found, add a tuple (offset, length, next char)
		if bestMatchLength > 0 && i+bestMatchLength < len(lookahead) {
			nextChar := lookahead[i+bestMatchLength]
			compressedData = append(compressedData, LZ77Tuple{bestMatchDistance, bestMatchLength, nextChar})
			i += bestMatchLength + 1
		} else {
			// No match found, add a tuple with zero offset and length, and the current character
			compressedData = append(compressedData, LZ77Tuple{0, 0, lookahead[i]})
			i++
		}

		// Update the history window
		if i < len(lookahead) {
			history += lookahead[:i]
			if len(history) > windowSize {
				history = history[len(history)-windowSize:]
			}
		}
	}

	return compressedData
}

// sliceAndCompress slices the input string, divides it, compresses it, and prints results
func sliceAndCompress(input string, n int) {
	// Ensure n does not exceed the length of input
	if n > len(input) {
		n = len(input)
	}

	newText := input[:n]
	fmt.Println("NewText:", newText)

	// Calculate the 2/3 and 1/3 split
	twoThirdsIndex := (2 * n) / 3
	history := newText[:twoThirdsIndex]
	lookahead := newText[twoThirdsIndex:]

	fmt.Println("History:", history)
	fmt.Println("Lookahead:", lookahead)

	// Apply LZ77 compression on history and lookahead
	compressedData := compressLZ77(history, lookahead)

	// Print compressed data
	fmt.Println("Compressed Data:")
	for _, tuple := range compressedData {
		fmt.Printf("(%d, %d, %q)\n", tuple.offset, tuple.length, tuple.char)
	}

	// Calculate memory size of compressed data
	memorySizeCompressed := 0
	for _, tuple := range compressedData {
		memorySizeCompressed += len(fmt.Sprintf("%d%d%c", tuple.offset, tuple.length, tuple.char))
	}
	memorySizeOriginal := len(newText)

	fmt.Println("Memory Size of Compressed Data:", memorySizeCompressed)
	fmt.Println("Memory Size of Original Data:", memorySizeOriginal)

	// Compression ratio
	compressionRatio := float64(memorySizeCompressed) / float64(memorySizeOriginal)
	fmt.Printf("Compression Ratio: %.2f\n", compressionRatio)
}

func main() {
	input := "In radio c'è un pulcino In radio c'è un pulcino È il pulcino pio, è il pulcino pio E il pulcino pio, e il pulcino pio"
	plaque := "2077518"
	n := 0
	for i := 0; i < len(plaque); i++ {
		n += int(plaque[i] - '0')
	}

	sliceAndCompress(input, n)
}
